sw/device/lib/runtime/ibex.h

```diff
@@ -75,4 +91,44 @@ uint32_t ibex_mcause_read(void);
  */
 uint32_t ibex_mtval_read(void);
 
+/**
+ * Initializes the ibex timeout based on current mcycle count.
+ *
+ * @param timeout_usec Timeout in microseconds.
+ * @return The initialized timeout value.
+ */
+inline ibex_timeout_t ibex_timeout_init(uint32_t timeout_usec) {
+  return (ibex_timeout_t){.cycles = kClockFreqCpuHz * timeout_usec / 1000000,
+                          .start = ibex_mcycle_read()};
+}
+
+/**
+ * Returns boolean indicating the timeout occurred.
+ *
+ * @param timeout Ibex timeout value
+ * @return Boolean indicating the timeout occurred.
+ */
+inline bool ibex_timeout_check(ibex_timeout_t *timeout) {
+  return ibex_mcycle_read() - timeout->start < timeout->cycles;
+}
+
+/**
+ * Convenience macro to spin with timeout in microseconds.
+ *
+ * @param expr An expression that is evaluated.
```

Comment:
Maybe it's worth noting that the expression will be evaluated multiple times?

Reply:
Right... most likely it will involve some function call, like a CSR read.